ee/fleetd-chrome/src/background.ts

```diff
@@ -0,0 +1,241 @@
+    try {
+      const query_result = await DATABASE.query(query_sql);
+      results[query_name] = query_result;
+      statuses[query_name] = 0;
+    } catch (err) {
+      console.warn(`Query (${query_name} sql: "${query_sql}") failed: ${err}`);
+      results[query_name] = null;
+      statuses[query_name] = 1;
+      messages[query_name] = err.toString();
+    }
+  }
+
+  const live_query_result_request = {
+    queries: results,
+    statuses,
+    messages,
+  };
+
+  await authenticatedRequest({
+    path: "/api/osquery/distributed/write",
+    body: live_query_result_request,
+  });
+};
+
+const getNodeKey = async () => {
+  const { node_key } = await chrome.storage.local.get("node_key");
+  return node_key;
+};
+
+const clearNodeKey = async () => {
+  await chrome.storage.local.remove("node_key");
+};
+
+const setNodeKey = async (node_key: string) => {
+  await chrome.storage.local.set({ node_key });
+};
+
+const main = async () => {
+  console.debug("main");
+
+  // @ts-expect-error @types/chrome doesn't yet have navigator.userAgentData.
+  const platform = navigator.userAgentData.platform;
+  const { installType } = await chrome.management.getSelf();
+  if (platform !== "Chrome OS" && installType !== "development") {
+    console.error("Refusing to run on non Chrome OS with managed install!");
+    return;
+  }
+
+  if (!DATABASE) {
+    const module = await SQLiteAsyncESMFactory();
+    const sqlite3 = SQLite.Factory(module);
+    const db = await sqlite3.open_v2(":memory:");
+
+    const virtual = new VirtualDatabase(sqlite3, db);
+    DATABASE = virtual;
+
+    // Expose it for debugging in console
+    globalThis.DB = DATABASE;
+  }
+
+  const node_key = await getNodeKey();
+  if (!node_key) {
+    await enroll();
+  }
+  await live_query();
+  //await sqlite3.close(db);
+};
+
+class NodeInvalidError extends Error {
+  constructor(message: string) {
+    super(`request failed with node_invalid: ${message}`);
+    this.name = "NodeInvalidError";
+  }
+}
+
+// QUESTION maybe we should use one of the persistence mechanisms described in
+// https://stackoverflow.com/a/66618269/491710? The "offscreen API" mechanism might be useful. On
+// the other hand, this seems to work decently well and adding the complexity might not be worth it.
+
+// This is a bit funky here. We want the main loop to run every 10 seconds, but we have to be
+// careful that we clear the old timeouts because of the alarm triggering that causes an additional
+// call to mainLoop. If we don't clear the timeout, we'll start getting more and more calls to
+// mainLoop each time the alarm fires.
+let mainTimeout: ReturnType<typeof setTimeout>;
+const mainLoop = async () => {
+  await main();
+  clearTimeout(mainTimeout);
+  mainTimeout = setTimeout(mainLoop, 10 * 1000);
+};
+mainLoop();
+
+// This alarm is used to ensure the extension "wakes up" at least once every minute. Otherwise
+// Chrome could shut it down in the background.
```

Comment:
What events trigger Chrome to load/reload the service worker? Is that something that can be configured?

Reply:
My understanding is that Chrome does it automatically and it is not configurable. Seems to be something that is done to save system resources.